lisa/tests/kernel/scheduler/load_tracking.py

```diff
@@ -733,3 +735,218 @@ def test_load_avg_behaviour(self, error_margin_pct=5, allowed_error_pct=5)\
+    sum of the utilization of all the tasks running on it.
+
+    **Design notes:**
+
+    Since we sum up the utilization of each task, make sure not to overload the
+    CPU - IOW, there should always be some idle cycles.
+
+    The code assumes all tasks have the same number of phases, and that those
+    phases are all aligned.
+    """
+
+    PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
+    """
+    The duration of a single phase
+    """
+
+    @classmethod
+    def _run_rtapp(cls, te, res_dir, profile):
+        # Just do some validation on the profile
+        for name, task in profile.items():
+            for phase in task.phases:
+                if len(phase.cpus) != 1 :
+                    raise RuntimeError("Each phase must be tied to a single CPU. "
+                                       "Task \"{}\" violates this".format(name))
+
+        super()._run_rtapp(te, res_dir, profile)
+
+    def __init__(self, res_dir, plat_info, rtapp_profile):
+        super().__init__(res_dir, plat_info, rtapp_profile)
+
+        self.cpus = set()
+
+        self.reference_task = list(self.rtapp_profile.values())[0]
+        self.nr_phases = len(self.reference_task.phases)
+
+        for task in self.rtapp_profile.values():
+            for phase in task.phases:
+                self.cpus.update(phase.cpus)
+
+        self.phases_durations = [phase.duration_s
+                                 for phase in self.reference_task.phases]
+
+    def get_expected_cpu_util(self):
+        """
+        Get the per-phase average CPU utilization expected from the rtapp profile
+
+        :returns: A dict of the shape {cpu : {phase_id : expected_util}}
+        """
+        cpu_util = {cpu : {phase_id : 0 for phase_id in range(self.nr_phases)}
+                    for cpu in self.cpus}
+
+        for task in self.rtapp_profile.values():
+            for phase_id, phase in enumerate(task.phases):
+                cpu_util[phase.cpus[0]][phase_id] += UTIL_SCALE * (phase.duty_cycle_pct / 100)
+
+        return cpu_util
+
+    def get_trace_cpu_util(self):
+        """
+        Get the per-phase average CPU utilization read from the trace
+
+        :returns: A dict of the shape {cpu : {phase_id : trace_util}}
+        """
+        cpu_util = {cpu : {phase_id : 0 for phase_id in range(self.nr_phases)}
+                    for cpu in self.cpus}
+        df = self.trace.analysis.load_tracking.df_cpus_signals()
+        sw_df = self.trace.df_events("sched_switch")
+
+        phase_start = sw_df[sw_df.next_comm == list(self.rtapp_profile.keys())[0]].index[0]
+
+        for phase in range(self.nr_phases):
+            # Start looking at signals once they should've converged
+            start = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
+            # Trim the end a bit, otherwise we could have one or two events
+            # from the next phase
+            end = phase_start + self.phases_durations[phase] * .9
```

Comment:
Since we have:
```python
PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
```
I would say it could be just good to set:
```python
end = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
```

Reply:
Actually, giving the following code, where you keep computing (end-start), why not computing directly here:
```python
stable_interval = end - start
```
and use it in L826 ?

Reply:
> Since we have:
> 
> ```python
> PHASE_DURATION_S = 3 * UTIL_AVG_CONVERGENCE_TIME_S
> ```
> 
> I would say it could be just good to set:
> 
> ```python
> end = phase_start + UTIL_AVG_CONVERGENCE_TIME_S
> ```

IMO the larger the interval is, the better it is in terms of averages.



> Actually, giving the following code, where you keep computing (end-start), why not computing directly here:
> 
> ```python
> stable_interval = end - start
> ```
> 
> and use it in L826 ?

Can do.